core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryJoinOptimizer.java

```diff
@@ -109,7 +109,9 @@ public void meet(Join node) {
 					Map<TupleExpr, List<Var>> varsMap = new HashMap<>();
 
 					for (TupleExpr tupleExpr : joinArgs) {
-						cardinalityMap.put(tupleExpr, statistics.getCardinality(tupleExpr));
+						double cardinality = statistics.getCardinality(tupleExpr);
+						tupleExpr.setEstimatedRows(cardinality);
```

Comment:
Right, so here's where I went wrong: I somehow had it in my had that you were using the outcome of `QueryJoinOptimizer.getTupleExprCardinality` to set this number. Apologies for the confusion.

Btw that might be a good thing to fix as part of this PR: rename that method to `getTupleExprCost`.